File: src/main/java/com/kkllffaa/meteor_litematica_printer/Modules/AutoTool.java
package com.kkllffaa.meteor_litematica_printer.Modules;

import meteordevelopment.meteorclient.events.entity.player.StartBreakingBlockEvent;
import meteordevelopment.meteorclient.events.world.TickEvent;
import meteordevelopment.meteorclient.settings.*;
import meteordevelopment.meteorclient.systems.modules.Module;
import meteordevelopment.meteorclient.systems.modules.Modules;
import meteordevelopment.meteorclient.systems.modules.render.Xray;
import meteordevelopment.meteorclient.systems.modules.world.InfinityMiner;
import meteordevelopment.meteorclient.utils.Utils;
import meteordevelopment.meteorclient.utils.player.FindItemResult;
import meteordevelopment.meteorclient.utils.player.InvUtils;
import meteordevelopment.meteorclient.utils.world.BlockUtils;
import meteordevelopment.orbit.EventHandler;
import meteordevelopment.orbit.EventPriority;
import net.minecraft.block.*;
import net.minecraft.component.DataComponentTypes;
import net.minecraft.enchantment.Enchantments;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.ShearsItem;
import net.minecraft.registry.tag.BlockTags;
import net.minecraft.registry.tag.ItemTags;

import java.util.List;
import java.util.function.Predicate;

import com.kkllffaa.meteor_litematica_printer.Addon;

public class AutoTool extends Module {
    private final SettingGroup sgGeneral = settings.getDefaultGroup();
    private final SettingGroup sgWhitelist = settings.createGroup("Whitelist");

    //region General

    private final Setting<EnchantPreference> prefer = sgGeneral.add(new EnumSetting.Builder<EnchantPreference>()
        .name("prefer")
        .description("Either to prefer Silk Touch, Fortune, or none.")
        .defaultValue(EnchantPreference.SilkTouch)
        .build()
    );

    private final Setting<Boolean> silkTouchForEnderChest = sgGeneral.add(new BoolSetting.Builder()
        .name("silk-touch-for-ender-chest")
        .description("Mines Ender Chests only with the Silk Touch enchantment.")
        .defaultValue(true)
        .build()
    );

    private final Setting<Boolean> fortuneForOresCrops = sgGeneral.add(new BoolSetting.Builder()
        .name("fortune-for-ores-and-crops")
        .description("Mines Ores and crops only with the Fortune enchantment.")
        .defaultValue(false)
        .build()
    );

    private final Setting<Boolean> antiBreak = sgGeneral.add(new BoolSetting.Builder()
        .name("anti-break")
        .description("Stops you from breaking your tool.")
        .defaultValue(true)
        .build()
    );

    private final Setting<Integer> breakDurability = sgGeneral.add(new IntSetting.Builder()
        .name("anti-break-percentage")
        .description("The durability percentage to stop using a tool.")
        .defaultValue(9)
        .range(1, 100)
        .sliderRange(1, 100)
        .visible(antiBreak::get)
        .build()
    );

    private final Setting<Integer> switchBackDelay = sgGeneral.add((new IntSetting.Builder()
        .name("switch-back-delay")
        .description("Delay in ticks for switching tools back.")
        .defaultValue(5)
        .build()
    ));

    private final Setting<Integer> useSlot = sgGeneral.add(new IntSetting.Builder()
        .name("use-slot")
        .description("The static hotbar slot to use.")
        .defaultValue(2)
        .range(1, 9)
        .sliderRange(1, 9)
        .build()
    ); 
    //endregion

    //region Whitelist and blacklist

    private final Setting<ListMode> listMode = sgWhitelist.add(new EnumSetting.Builder<ListMode>()
        .name("list-mode")
        .description("Selection mode.")
        .defaultValue(ListMode.Blacklist)
        .build()
    );

    private final Setting<List<Item>> whitelist = sgWhitelist.add(new ItemListSetting.Builder()
        .name("whitelist")
        .description("The tools you want to use.")
        .visible(() -> listMode.get() == ListMode.Whitelist)
        .filter(AutoTool::isTool)
        .build()
    );

    private final Setting<List<Item>> blacklist = sgWhitelist.add(new ItemListSetting.Builder()
        .name("blacklist")
        .description("The tools you don't want to use.")
        .visible(() -> listMode.get() == ListMode.Blacklist)
        .filter(AutoTool::isTool)
        .build()
    );
    //endregion

    private int tick = -1;

    public AutoTool() {
        super(Addon.TOOLSCATEGORY, "auto-tool-in-one-slot", "Automatically switches to the most effective tool when performing an action.");
    }
    @Override
    public void onActivate() {
        resolveModuleConflict();
    }
    private void resolveModuleConflict() {
        Module meteorAutoTool = Modules.get().get(meteordevelopment.meteorclient.systems.modules.player.AutoTool.class);
        if (meteorAutoTool != null && meteorAutoTool.isActive()) {
            meteorAutoTool.toggle();
        }
    }

    @EventHandler
    private void onTick(TickEvent.Post event) {
        if (Modules.get().isActive(InfinityMiner.class)) return;

        if (tick > 0) {
            tick--;
        }
        if (mc.options.attackKey.isPressed()) {
            tick = switchBackDelay.get();
        }
        if (tick == 0) {
            InvUtils.swapBack();
            tick = -1;
        }

    }

    @EventHandler(priority = EventPriority.HIGH)
    private void onStartBreakingBlock(StartBreakingBlockEvent event) {
        if (Modules.get().isActive(InfinityMiner.class)) return;
        if (mc.player.isCreative()) return;

        // Get blockState
        BlockState blockState = mc.world.getBlockState(event.blockPos);
        if (!BlockUtils.canBreak(event.blockPos, blockState)) return;

        // Check if we should switch to a better tool
        int useSlotIndex = useSlot.get() - 1;
        ItemStack useSlotStack = mc.player.getInventory().getStack(useSlotIndex);

        double bestScore = -1;
        int bestSlot = -1;

        for (int i = 0; i < 36; i++) {
            ItemStack itemStack = mc.player.getInventory().getStack(i);

            if (listMode.get() == ListMode.Whitelist && !whitelist.get().contains(itemStack.getItem())) continue;
            if (listMode.get() == ListMode.Blacklist && blacklist.get().contains(itemStack.getItem())) continue;

            double score = getScore(itemStack, blockState, silkTouchForEnderChest.get(), fortuneForOresCrops.get(), prefer.get(), itemStack2 -> !shouldStopUsing(itemStack2));
            if (score < 0) continue;

            if (score > bestScore) {
                bestScore = score;
                bestSlot = i;
            }
        }

        if (bestSlot != -1 && (bestScore > getScore(useSlotStack, blockState, silkTouchForEnderChest.get(), fortuneForOresCrops.get(), prefer.get(), itemStack -> !shouldStopUsing(itemStack)) || shouldStopUsing(useSlotStack) || !isTool(useSlotStack))) {
            if (bestSlot != useSlotIndex) {
                InvUtils.move().fromHotbar(bestSlot).to(useSlotIndex);
                if (!mc.player.currentScreenHandler.getCursorStack().isEmpty()) {
                    FindItemResult emptySlot = InvUtils.findEmpty();
                    if (emptySlot.found()) {
                        InvUtils.click().slot(emptySlot.slot());
                    } else {
                        InvUtils.click().slot(bestSlot);
                        info("No empty slot found, put back to original slot");
                    }
                }
            }
        }
        
        InvUtils.swap(useSlotIndex, true);
        
        // Anti break
        ItemStack currentStack = mc.player.getMainHandStack();

        if (shouldStopUsing(currentStack) && isTool(currentStack)) {
            mc.options.attackKey.setPressed(false);
            event.cancel();
        }
    }

    private boolean shouldStopUsing(ItemStack itemStack) {
        return antiBreak.get() && (itemStack.getMaxDamage() - itemStack.getDamage()) < (itemStack.getMaxDamage() * breakDurability.get() / 100);
    }

    public static double getScore(ItemStack itemStack, BlockState state, boolean silkTouchEnderChest, boolean fortuneOre, EnchantPreference enchantPreference, Predicate<ItemStack> good) {
        if (!good.test(itemStack) || !isTool(itemStack)) return -1;
        if (!itemStack.isSuitableFor(state) &&
            !(itemStack.isIn(ItemTags.SWORDS) && (state.getBlock() instanceof BambooBlock || state.getBlock() instanceof BambooShootBlock)) &&
            !(itemStack.getItem() instanceof ShearsItem && state.getBlock() instanceof LeavesBlock || state.isIn(BlockTags.WOOL)))
            return -1;

        if (silkTouchEnderChest
            && state.getBlock() == Blocks.ENDER_CHEST
            && !Utils.hasEnchantments(itemStack, Enchantments.SILK_TOUCH)) {
            return -1;
        }

        if (fortuneOre
            && isFortunable(state.getBlock())
            && !Utils.hasEnchantments(itemStack, Enchantments.FORTUNE)) {
            return -1;
        }

        double score = 0;

        score += itemStack.getMiningSpeedMultiplier(state) * 1000;
        score += Utils.getEnchantmentLevel(itemStack, Enchantments.UNBREAKING);
        score += Utils.getEnchantmentLevel(itemStack, Enchantments.EFFICIENCY);
        score += Utils.getEnchantmentLevel(itemStack, Enchantments.MENDING);

        if (enchantPreference == EnchantPreference.Fortune) score += Utils.getEnchantmentLevel(itemStack, Enchantments.FORTUNE);
        if (enchantPreference == EnchantPreference.SilkTouch) score += Utils.getEnchantmentLevel(itemStack, Enchantments.SILK_TOUCH);

        if (itemStack.isIn(ItemTags.SWORDS) && (state.getBlock() instanceof BambooBlock || state.getBlock() instanceof BambooShootBlock))
            score += 9000 + (itemStack.get(DataComponentTypes.TOOL).getSpeed(state) * 1000);

        return score;
    }

    public static boolean isTool(Item item) {
        return isTool(item.getDefaultStack());
    }

    public static boolean isTool(ItemStack itemStack) {
        return itemStack.isIn(ItemTags.AXES) || itemStack.isIn(ItemTags.HOES) || itemStack.isIn(ItemTags.PICKAXES) || itemStack.isIn(ItemTags.SHOVELS) || itemStack.getItem() instanceof ShearsItem;
    }

    private static boolean isFortunable(Block block) {
        if (block == Blocks.ANCIENT_DEBRIS) return false;
        return Xray.ORES.contains(block) || block instanceof CropBlock;
    }

    public enum EnchantPreference {
        None,
        Fortune,
        SilkTouch
    }

    public enum ListMode {
        Whitelist,
        Blacklist
    }
}

File: src/main/kotlin/com/kkllffaa/meteor_litematica_printer/Modules/Tools/AutoTool.kt
package com.kkllffaa.meteor_litematica_printer.Modules.Tools

import com.kkllffaa.meteor_litematica_printer.Addon
import meteordevelopment.meteorclient.events.entity.player.StartBreakingBlockEvent
import meteordevelopment.meteorclient.events.world.TickEvent
import meteordevelopment.meteorclient.settings.*
import meteordevelopment.meteorclient.systems.modules.Module
import meteordevelopment.meteorclient.systems.modules.Modules
import meteordevelopment.meteorclient.systems.modules.player.AutoTool
import meteordevelopment.meteorclient.systems.modules.render.Xray
import meteordevelopment.meteorclient.systems.modules.world.InfinityMiner
import meteordevelopment.meteorclient.utils.Utils
import meteordevelopment.meteorclient.utils.player.InvUtils
import meteordevelopment.meteorclient.utils.player.SlotUtils
import meteordevelopment.meteorclient.utils.world.BlockUtils
import meteordevelopment.orbit.EventHandler
import meteordevelopment.orbit.EventPriority
import net.minecraft.block.*
import net.minecraft.component.DataComponentTypes
import net.minecraft.component.type.ToolComponent
import net.minecraft.enchantment.Enchantments
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.item.ShearsItem
import net.minecraft.registry.tag.BlockTags
import net.minecraft.registry.tag.ItemTags
import java.util.function.Predicate

class AutoTool :
    Module(Addon.TOOLS, "auto-tool-+", "Automatically switches to the most effective tool when performing an action.") {
    private val sgGeneral: SettingGroup = settings.getDefaultGroup()
    private val sgWhitelist: SettingGroup = settings.createGroup("Whitelist")

    //region General
    private val prefer: Setting<EnchantPreference?> = sgGeneral.add<EnchantPreference?>(
        EnumSetting.Builder<EnchantPreference?>()
            .name("prefer")
            .description("Either to prefer Silk Touch, Fortune, or none.")
            .defaultValue(EnchantPreference.SilkTouch)
            .build()
    )

    private val silkTouchForEnderChest: Setting<Boolean?> = sgGeneral.add<Boolean?>(
        BoolSetting.Builder()
            .name("silk-touch-for-ender-chest")
            .description("Mines Ender Chests only with the Silk Touch enchantment.")
            .defaultValue(true)
            .build()
    )

    private val fortuneForOresCrops: Setting<Boolean?> = sgGeneral.add<Boolean?>(
        BoolSetting.Builder()
            .name("fortune-for-ores-and-crops")
            .description("Mines Ores and crops only with the Fortune enchantment.")
            .defaultValue(false)
            .build()
    )

    private val antiBreak: Setting<Boolean?> = sgGeneral.add<Boolean?>(
        BoolSetting.Builder()
            .name("anti-break")
            .description("Stops you from breaking your tool.")
            .defaultValue(true)
            .build()
    )

    private val breakDurability: Setting<Int?> = sgGeneral.add<Int?>(
        IntSetting.Builder()
            .name("anti-break-percentage")
            .description("The durability percentage to stop using a tool.")
            .defaultValue(9)
            .range(1, 100)
            .sliderRange(1, 100)
            .visible(IVisible { antiBreak.get()!! })
            .build()
    )

    private val switchBackDelay: Setting<Int?> = sgGeneral.add<Int?>(
        (IntSetting.Builder()
            .name("switch-back-delay")
            .description("Delay in ticks for switching tools back.")
            .defaultValue(15)
            .range(0, 100)
            .build()
                )
    )

    private val useSlot: Setting<Int?> = sgGeneral.add<Int?>(
        IntSetting.Builder()
            .name("use-slot")
            .description("the only one static hotbar slot to use.")
            .defaultValue(2)
            .range(1, 9)
            .sliderRange(1, 9)
            .build()
    )

    //endregion
    //region Whitelist and blacklist
    private val listMode: Setting<ListMode?> = sgWhitelist.add<ListMode?>(
        EnumSetting.Builder<ListMode?>()
            .name("list-mode")
            .description("Selection mode.")
            .defaultValue(ListMode.Blacklist)
            .build()
    )

    private val whitelist: Setting<MutableList<Item?>?> = sgWhitelist.add<MutableList<Item?>?>(
        ItemListSetting.Builder()
            .name("whitelist")
            .description("The tools you want to use.")
            .visible(IVisible { listMode.get() == ListMode.Whitelist })
            .filter(Predicate { item: Item? -> Companion.isTool(item!!) })
            .build()
    )

    private val blacklist: Setting<MutableList<Item?>?> = sgWhitelist.add<MutableList<Item?>?>(
        ItemListSetting.Builder()
            .name("blacklist")
            .description("The tools you don't want to use.")
            .visible(IVisible { listMode.get() == ListMode.Blacklist })
            .filter(Predicate { item: Item? -> Companion.isTool(item!!) })
            .build()
    )

    private var useSlotIndex = useSlot.get()!! - 1

    override fun onActivate() {
        resolveModuleConflict()
    }

    private fun resolveModuleConflict() {
        val meteorAutoTool: Module? = Modules.get().get<AutoTool?>(AutoTool::class.java)
        if (meteorAutoTool != null && meteorAutoTool.isActive()) {
            meteorAutoTool.toggle()
        }
    }

    @EventHandler
    private fun onTick(event: TickEvent.Post?) {
        if (Modules.get().isActive(InfinityMiner::class.java)) return

        if (mc.interactionManager != null && mc.interactionManager!!.isBreakingBlock()) {
            busyTick = switchBackDelay.get()!!
        }
        if (busyTick == 0) {
            if (useSlotIndex == mc.player!!.getInventory().getSelectedSlot()) {
                InvUtils.swapBack()
            } else {
                InvUtils.previousSlot = -1
            }
            busyTick = -1
        }
        if (busyTick > 0) {
            busyTick--
        }
    }

    @EventHandler(priority = EventPriority.HIGH)
    private fun onStartBreakingBlock(event: StartBreakingBlockEvent) {
        if (Modules.get().isActive(InfinityMiner::class.java)) return
        if (mc.player!!.isCreative()) return

        val blockState = mc.world!!.getBlockState(event.blockPos)
        if (!BlockUtils.canBreak(event.blockPos, blockState)) return

        // Check if we should switch to a better tool
        var bestScore = -1.0
        var bestSlot = -1

        for (i in 0..35) {
            val itemStack = mc.player!!.getInventory().getStack(i)

            if (listMode.get() == ListMode.Whitelist && !whitelist.get()!!.contains(itemStack.getItem())) continue
            if (listMode.get() == ListMode.Blacklist && blacklist.get()!!.contains(itemStack.getItem())) continue

            val score: Double = Companion.getScore(
                itemStack,
                blockState,
                silkTouchForEnderChest.get()!!,
                fortuneForOresCrops.get()!!,
                prefer.get(),
                Predicate { itemStack2: ItemStack? -> !shouldStopUsing(itemStack2!!) })

            if (score > bestScore) {
                bestScore = score
                bestSlot = i
            }
        }
        if (bestSlot == -1) {
            val cursorStack = mc.player!!.currentScreenHandler.getCursorStack()
            if (!cursorStack.isEmpty() && !(listMode.get() == ListMode.Whitelist && !whitelist.get()!!
                    .contains(cursorStack.getItem())) && !(listMode.get() == ListMode.Blacklist && blacklist.get()!!
                    .contains(cursorStack.getItem()))
            ) {
                val score: Double = Companion.getScore(
                    cursorStack,
                    blockState,
                    silkTouchForEnderChest.get()!!,
                    fortuneForOresCrops.get()!!,
                    prefer.get(),
                    Predicate { itemStack2: ItemStack? -> !shouldStopUsing(itemStack2!!) })
                if (score > bestScore) {
                    bestScore = score
                    bestSlot = -2
                }
            }
        }

        if (bestSlot != -1 && bestScore > Companion.getScore(
                mc.player!!.getMainHandStack(),
                blockState,
                silkTouchForEnderChest.get()!!,
                fortuneForOresCrops.get()!!,
                prefer.get(),
                Predicate { itemStack: ItemStack? -> !shouldStopUsing(itemStack!!) })
        ) {
            if (bestSlot != mc.player!!.getInventory().getSelectedSlot()) {
                if (SlotUtils.isHotbar(bestSlot)) {
                    useSlotIndex = bestSlot
                } else {
                    useSlotIndex = useSlot.get()!! - 1
                    if (bestSlot == -2) {
                        InvUtils.click().slot(useSlotIndex)
                        if (!mc.player!!.currentScreenHandler.getCursorStack().isEmpty()) {
                            val emptySlot = InvUtils.findEmpty()
                            if (emptySlot.found()) {
                                InvUtils.click().slot(emptySlot.slot())
                            } else {
                                warning("No empty slot found")
                            }
                        }
                    } else {
                        InvUtils.move().fromHotbar(bestSlot).to(useSlotIndex)
                        if (!mc.player!!.currentScreenHandler.getCursorStack().isEmpty()) {
                            val emptySlot = InvUtils.findEmpty()
                            if (emptySlot.found()) {
                                InvUtils.click().slot(emptySlot.slot())
                            } else {
                                InvUtils.click().slot(bestSlot)
                                warning("No empty slot found, put back to original slot")
                            }
                        }
                    }
                }
                InvUtils.swap(useSlotIndex, true)
            }
        } else {
            //没有有耐久的工具
        }
        // Anti break
        val currentStack = mc.player!!.getMainHandStack()

        if (shouldStopUsing(currentStack) && isTool(currentStack)) {
            mc.options.attackKey.setPressed(false)
            event.cancel()
        } else {
            busyTick = switchBackDelay.get()!!
        }
    }

    private fun shouldStopUsing(itemStack: ItemStack): Boolean {
        return antiBreak.get() && (itemStack.getMaxDamage() - itemStack.getDamage()) < (itemStack.getMaxDamage() * breakDurability.get()!! / 100)
    }

    enum class EnchantPreference {
        None,
        Fortune,
        SilkTouch
    }

    enum class ListMode {
        Whitelist,
        Blacklist
    }

    companion object {
        //endregion
        var busyTick: Int = -1
        fun getScore(
            itemStack: ItemStack,
            state: BlockState,
            silkTouchEnderChest: Boolean,
            fortuneOre: Boolean,
            enchantPreference: EnchantPreference?,
            good: Predicate<ItemStack?>
        ): Double {
            if (!good.test(itemStack) || !isTool(itemStack)) return -1.0
            if (!itemStack.isSuitableFor(state) && !(itemStack.isIn(ItemTags.SWORDS) && (state.getBlock() is BambooBlock || state.getBlock() is BambooShootBlock)) && !(itemStack.getItem() is ShearsItem && state.getBlock() is LeavesBlock || state.isIn(
                    BlockTags.WOOL
                ))
            ) return -1.0

            if (silkTouchEnderChest
                && state.getBlock() === Blocks.ENDER_CHEST && !Utils.hasEnchantments(itemStack, Enchantments.SILK_TOUCH)
            ) {
                return -1.0
            }

            if (fortuneOre
                && isFortunable(state.getBlock())
                && !Utils.hasEnchantments(itemStack, Enchantments.FORTUNE)
            ) {
                return -1.0
            }

            var score = 0.0

            score += (itemStack.getMiningSpeedMultiplier(state) * 1000).toDouble()
            score += Utils.getEnchantmentLevel(itemStack, Enchantments.UNBREAKING).toDouble()
            score += Utils.getEnchantmentLevel(itemStack, Enchantments.EFFICIENCY).toDouble()
            score += Utils.getEnchantmentLevel(itemStack, Enchantments.MENDING).toDouble()

            if (enchantPreference == EnchantPreference.Fortune) score += Utils.getEnchantmentLevel(
                itemStack,
                Enchantments.FORTUNE
            ).toDouble()
            if (enchantPreference == EnchantPreference.SilkTouch) score += Utils.getEnchantmentLevel(
                itemStack,
                Enchantments.SILK_TOUCH
            ).toDouble()

            if (itemStack.isIn(ItemTags.SWORDS) && (state.getBlock() is BambooBlock || state.getBlock() is BambooShootBlock)) score += (9000 + (itemStack.get<ToolComponent?>(
                DataComponentTypes.TOOL
            )!!.getSpeed(state) * 1000)).toDouble()

            return score
        }

        fun isTool(item: Item): Boolean {
            return isTool(item.getDefaultStack())
        }

        fun isTool(itemStack: ItemStack): Boolean {
            return itemStack.isIn(ItemTags.AXES) || itemStack.isIn(ItemTags.HOES) || itemStack.isIn(ItemTags.PICKAXES) || itemStack.isIn(
                ItemTags.SHOVELS
            ) || itemStack.getItem() is ShearsItem
        }

        private fun isFortunable(block: Block?): Boolean {
            if (block === Blocks.ANCIENT_DEBRIS) return false
            return Xray.ORES.contains(block) || block is CropBlock
        }
    }
}
